Adopt aggregated apiserver

aylei opened the pull request:
Signed-off-by: Aylei <rayingecho@gmail.com>

### What problem does this PR solve? <!--add and issue link with summary if exists-->
#1002 

### What is changed and how does it work?
Add aggregated apiserver skeleton and toolchain.

More todos are listed in the original issue #1002 

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

Manual test

- [x] Create
- [x] Update
- [x] List
- [x] Watch

Code changes

 - Has Go code change

### Does this PR introduce a user-facing change?:
 <!--
 If no, just write "NONE" in the release-note block below.
 If yes, a release note is required:
 Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
 -->
 ```release-note
New component TiDB aggregated apiserver.
 ```

Here's an example about adding new resources in AA: https://github.com/aylei/tidb-operator/pull/1

To make review easier, I split this PR to multiple commits:

- Pin the versions of dependencies for Aggregated Apiserver, which actually can be ignored: https://github.com/pingcap/tidb-operator/pull/1048/commits/9103ca8f36e59c96f0ae5e3e00d2407a718d2012
- Add dataresource CRD, which introduce a new CRD called `DataResource` as the storage of our AA: https://github.com/pingcap/tidb-operator/pull/1048/commits/9103ca8f36e59c96f0ae5e3e00d2407a718d2012
- Generate code for dataresouce CRD, which can be ignored: https://github.com/pingcap/tidb-operator/pull/1048/commits/92fad18de2e4ce3d1806ce9f78bc898c5799dcf3
- **Important: Implement a kube-apiserver based `storage.Interface`**: https://github.com/pingcap/tidb-operator/pull/1048/commits/209a6e1ca04e23b3c9802c323653f97f054db58e
- Add aggregated apiserver skeleton and toolchain: https://github.com/pingcap/tidb-operator/pull/1048/commits/0fa2977740ba9993b9cd43405a2592ca0fa2a7d7

PTAL @weekface @cofyc @tennix @onlymellb @Yisaer @DanielZhangQD

aylei commented on docs/adding-resource-in-aa.md:
```diff
@@ -0,0 +1,90 @@
+# Adding New Resources in TiDB Aggregated Apiserver
```
A brief introduction of the usage

aylei commented on pkg/apiserver/cmd/start.go:
```diff
@@ -0,0 +1,309 @@
+}
+
+// StartApiServer starts an apiserver.
+func StartApiServer(
```
AA entrypoint

aylei commented on pkg/apiserver/cmd/start.go:
```diff
@@ -0,0 +1,309 @@
+
+// Config create an apiserver config
+func (o ServerOptions) Config() (*apiserver.Config, error) {
+	if err := o.RecommendedOptions.SecureServing.MaybeDefaultWithSelfSignedCerts(
```
The CA issue is left for later resolution

cofyc replied:
self-sign certs and skip tls verify should be enough for internal use

aylei commented:
/run-e2e-in-kind

aylei commented:
e2e is not working for this issue now, because the apiserver lib introduce a dependency `bitbucket.org/ww/goautoneg` that have to be downloaded thru hg, which is not installed in the CI env.

```
$ go mod why bitbucket.org/ww/goautoneg
# bitbucket.org/ww/goautoneg
github.com/pingcap/tidb-operator/pkg/apiserver/cmd
k8s.io/apiserver/pkg/endpoints/filters
k8s.io/apiserver/pkg/endpoints/handlers/responsewriters
k8s.io/apiserver/pkg/endpoints/handlers/negotiation
bitbucket.org/ww/goautoneg
```

I will fix it soon, tt does not block reviewing, though.

cofyc commented:
/run-e2e-in-kind

cofyc commented:
/run-e2e-in-kind

cofyc commented:
> e2e is not working for this issue now, because the apiserver lib introduce a dependency bitbucket.org/ww/goautoneg that have to be downloaded thru hg, which is not installed in the CI env.
I've added a command to install hg in CI script.

aylei commented:
@cofyc Thanks!

cofyc commented on tools/tools.go:
```diff
@@ -32,4 +32,5 @@ import (
 	_ "k8s.io/code-generator/cmd/openapi-gen"
 	_ "k8s.io/code-generator/cmd/register-gen"
 	_ "k8s.io/code-generator/cmd/set-gen"
+	_ "github.com/kubernetes-incubator/apiserver-builder-alpha/cmd/apiregister-gen"
```
why not use `sigs.k8s.io/apiserver-builder-alpha/cmd/apiregister-gen`?

this process is moved to https://github.com/kubernetes-sigs/apiserver-builder-alpha/  now.

aylei replied:
The pinned version v1.12 was still using the `kubernetes-incubator` import path, we can use `sigs.k8s.io....` after we bump the kubernetes dependencies to `v1.16`

cofyc commented:
because I renamed `pkg/apis/pingcap.com` to `pkg/apis/pingcap`, some conflicts must be resolved. regenerating code and updating imports should be enough

aylei commented:
@cofyc Thanks for noting!

aylei commented:
/run-e2e-in-kind

tennix commented on pkg/pdapi/config.go:
```diff
@@ -0,0 +1,210 @@
+// type definition from PD and get rid of the dependency of pd/server
+// TODO: use type definition from pd when pd separate the config to proper package
+
+// Copied from pd/server/config.go
```
Why copy this and is it for configuration management?

tennix replied:
Oh, get it from the comment above.

cofyc commented on hack/generate-internal-groups.sh:
```diff
@@ -0,0 +1,141 @@
+if [ "${GENS}" = "all" ] || grep -qw "registry" <<<"${GENS}"; then
+  echo "Generating api registries"
+  "${GOPATH}/bin/apiregister-gen" --input-dirs "$(codegen::join , "${APIS_DIR[@]}")" "$@"
+fi
```
except this, do we have other changes?
when we can vendor code-generator root directory, I'd like to use scripts from code-generator.

aylei replied:
I change the client output package in case of collision with the generated code of CRD and change the package path from `a.b.c` to `a` so that the behavior is consistent with higher version of kubernetes.

I'd like to change this script to call the vendored script in the code-generator directory after we upgrade the k8s dependencies to `v1.16`

cofyc commented on pkg/apiserver/storage/store.go:
```diff
@@ -0,0 +1,464 @@
+)
+
+// store implements a ConfigMap backed storage.Interface
+type store struct {
```
add some unit tests to cover basic use cases?

cofyc commented on hack/aa-codegen.sh:
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash -e
```
`-e` flag is redundant, it equals to `set -o errexit`.

cofyc commented on hack/aa-codegen.sh:
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash -e
+
+# Copyright 2017 PingCAP, Inc.
```
nit: 2019

cofyc commented:
The whole idea looks good to me. Because it's a standalone apiserver and not used by other components. I think we can add some basic unit tests and merge it soon.

tennix commented on pkg/apiserver/storage/store.go:
```diff
@@ -0,0 +1,464 @@
+	}
+	v, err := conversion.EnforcePtr(listPtr)
+	if err != nil || v.Kind() != reflect.Slice {
+		panic("need ptr to slice")
```
return an error instead of just panic?

aylei replied:
This is the behavior of the official implementation, according to the context that this method get called, this condition indicates a bug in code and better to be treated as panic.

tennix replied:
OK, got it.

tennix commented on pkg/apiserver/storage/store.go:
```diff
@@ -0,0 +1,464 @@
+		return err
+	}
+	li, err := conversion.EnforcePtr(listPtr)
+	if err != nil || li.Kind() != reflect.Slice {
```
ditto

onlymellb commented on pkg/apiserver/cmd/start.go:
```diff
@@ -0,0 +1,335 @@
+		RunDelegatedAuth:   true,
+		Codec:              codec,
+	}
+	o.RecommendedOptions.SecureServing.BindPort = 443
```
Why not use [SecureServingOptions](https://github.com/kubernetes/apiserver/blob/fd43ea72cae9a713aae01016b7a58da1bbaf547a/pkg/server/options/serving.go#L37)  instead of a fixed secure port?  just like  `InsecureServingOptions` above.

onlymellb commented on pkg/apiserver/cmd/start.go:
```diff
@@ -0,0 +1,335 @@
+
+	loopbackKubeConfig, kubeInformerFactory, err := o.buildLoopback()
+	if err != nil {
+		klog.Warning("attempting to instantiate loopback client but failed...")
```
Add `err` message to the warning log.

aylei replied:
good catch

DanielZhangQD commented on pkg/apis/pingcap.com/v1alpha1/dataresource_types.go:
```diff
@@ -0,0 +1,41 @@
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Data []byte `json:"data"`
```
So for different CRs, they just store to the same structure []byte?
How to convert back to different CR structures from the Data?

aylei replied:
The data decode all information about the resource, just same as etcd

aylei commented:
/run-e2e-tests

aylei commented:
@cofyc @tennix @onlymellb @DanielZhangQD I've addressed the review comments and add the basic unit tests for `store.go`(https://github.com/pingcap/tidb-operator/pull/1048/commits/626f3f070c31d2367ad30bb8c80e26db26556bc3). Please take a look again and approve if there's no problem, thanks!

/run-e2e-in-kind

aylei commented:
/run-e2e-in-kind

aylei commented:
/run-e2e-in-kind

aylei commented:
/run-e2e-in-kind

cofyc commented:
/run-e2e-in-kind

cofyc commented:
/run-e2e-in-kind

aylei commented:
/run-e2e-in-kind
